Mathlib/Combinatorics/CombinatorialMap.lean

```diff
@@ -0,0 +1,187 @@
+/-
+Copyright (c) 2024 Kyle Miller, Rida Hamadani. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kyle Miller, Rida Hamadani
+-/
+import Mathlib.GroupTheory.Perm.Cycle.Basic
+
+/-!
+# Combinatorial Maps
+
+This module defines combinatorial maps, which are used to represent an object by its boundaries.
+
+Only two-dimensional combinatorial maps are considered.
+
+-/
+
+
+/--
+A two-dimension combinatorial map is a collection `D` of darts, and permutations `σ`, `α`, and `φ`.
+`α` is a fixed-point-free involution.
+
+Note: this definition excludes disjoint isolated vertices.
+-/
+structure CombinatorialMap (D : Type*) where
+  /--
+  Permutation `σ` whose orbits correspond to vertices. `σ` gives the next dart counter-clockwise
+  around an edge.
+  -/
+  σ : Equiv.Perm D
+  /-- Permutation `α` whose orbits correspond to edges. `α` gives the opposite dart of an edge. -/
+  α : Equiv.Perm D
+  /--
+  Permutation `φ` whose orbits correspond to faces. `φ` gives the next dart counter-clockwise of the
+  same face.
+  -/
+  φ : Equiv.Perm D
+  composition : φ * σ * α = 1
+  involutive : Function.Involutive α
+  fixedPoints_isEmpty : IsEmpty (Function.fixedPoints α)
+
+namespace CombinatorialMap
+
+variable {D D' : Type*} {M : CombinatorialMap D} {M' : CombinatorialMap D'}
+
+/-- The permutation `φ` expressed in terms of `σ` and `α`. -/
+lemma φ_eq : M.φ = M.α * M.σ⁻¹ := by
+  have h := M.composition
+  replace h := congr($h * M.α⁻¹ * M.σ⁻¹)
+  rw [mul_inv_cancel_right, mul_inv_cancel_right, one_mul] at h
+  rw [← M.involutive.symm_eq_self_of_involutive, h]
+  rfl
+
+lemma composition_apply {x : D} : M.φ (M.σ (M.α x)) = x := by
+  nth_rw 2 [← Equiv.Perm.one_apply x]
+  rw [← M.composition]
+  rfl
+
+/-- This order of permutations also results in the identity map. -/
+lemma composition' : M.α * M.φ * M.σ = 1 :=
+  have (y : D) : (M.α * M.φ * M.σ) y = y := by
+    obtain ⟨_, h⟩ := M.α.surjective y
+    simp [← h, composition_apply]
+  Equiv.Perm.ext fun _ ↦ this _
+
+lemma composition'_apply {x : D} : M.α (M.φ (M.σ x)) = x := by
+  convert_to (M.α * M.φ * M.σ) x = x
+  simp [M.composition']
+
+/--
+A homomorphism of `CombinatorialMap`s is a function that commutes with each of the respective
+permutations.
+-/
+def Hom (M : CombinatorialMap D) (M' : CombinatorialMap D') (f : D → D') : Prop :=
```

Comment:
This feels unidiomatic to me. Can we bundle the `f` into the structure, like we do for `SimpleGraph.Hom`?

We can define a separate `Iso` that is bijective.